File: Maximum Subarray Sum With Length Divisible by K/Maximum Subarray Sum With Length Divisible by K.cpp
// Solution for Maximum Subarray Sum With Length Divisible by K in CPP

class Solution {
public:
    long long maxSubarraySum(vector<int>& nums, int k) {

        int n = nums.size();

        // -----------------------------
        // Prefix Sum to compute any subarray sum in O(1)
        // prefixSum[i] = sum of nums[0..i]
        // -----------------------------
        vector<long long> prefixSum(n);
        prefixSum[0] = nums[0];

        for (int i = 1; i < n; i++) {
            prefixSum[i] = prefixSum[i - 1] + nums[i];
        }

        long long bestResult = LLONG_MIN;

        /*
            Key Observation:
            ----------------
            A subarray length must be divisible by k.
            That means valid subarrays start at indices:

                start = 0, 1, 2, ... , k-1

            Example for k = 3:
            start = 0 → take lengths 3, 6, 9 ...
            start = 1 → take lengths 3, 6, 9 ...
            start = 2 → take lengths 3, 6, 9 ...

            For each starting point, we repeatedly jump by k
            (i → i + k → i + 2k ...) and treat each group like
            a "Kadane's algorithm on chunks of size k".
        */

        for (int startIndex = 0; startIndex < k; startIndex++) {

            long long bestEndingHere = 0;  // Kadane's rolling sum
            int i = startIndex;

            while (i < n && (i + k - 1) < n) {

                // Ending index of this fixed-size subarray
                int endIndex = i + k - 1;

                // Compute subarray sum(nums[i..endIndex]) using prefix sums
                long long blockSum =
                    prefixSum[endIndex] - (i > 0 ? prefixSum[i - 1] : 0);

                /*
                    Update Kadane:
                    Either:
                     - Start a new subarray from this block
                     - Or extend previous sequence by adding this block
                */
                bestEndingHere = max(blockSum, bestEndingHere + blockSum);

                // Track global maximum
                bestResult = max(bestResult, bestEndingHere);

                // Move to next block of size k
                i += k;
            }
        }

        return bestResult;
    }
};

File: Maximum Subarray Sum With Length Divisible by K/Maximum Subarray Sum With Length Divisible by K.py
# Solution for Maximum Subarray Sum With Length Divisible by K in PY

class Solution:
    def maxSubarraySum(self, nums, k):
        n = len(nums)

        # Prefix sum
        prefix = [0] * n
        prefix[0] = nums[0]
        for i in range(1, n):
            prefix[i] = prefix[i - 1] + nums[i]

        best_result = float("-inf")

        # Try all starting indices modulo k
        for start_index in range(k):

            best_ending_here = 0   # Kadane on blocks
            i = start_index

            while i < n and (i + k - 1) < n:

                end_index = i + k - 1
                block_sum = prefix[end_index] - (prefix[i - 1] if i > 0 else 0)

                best_ending_here = max(block_sum, best_ending_here + block_sum)
                best_result = max(best_result, best_ending_here)

                i += k

        return best_result
